src/jace/stages.py

```diff
@@ -256,64 +279,130 @@ def as_sdfg(self) -> dace.SDFG:
+        flat_call_args = jax_tree.tree_leaves((args, kwargs))
+        flat_output = self._csdfg(flat_call_args)
+        if flat_output is None:
+            return None  # type: ignore[return-value]  # Type confusion.
+        return jax_tree.tree_unflatten(self._outtree, flat_output)
 
 
-#: Known compilation stages in JaCe.
-Stage = JaCeWrapped | JaCeLowered | JaCeCompiled
+# <--------------------------- Compilation/Optimization options management
+
+_JACELOWERED_ACTIVE_COMPILE_OPTIONS: CompilerOptions = optimization.DEFAULT_OPTIMIZATIONS.copy()
+"""Global set of currently active compilation/optimization options.
+
+The global set is initialized with `jace.optimization.DEFAULT_OPTIMIZATIONS`. It can be
+managed through `update_active_compiler_options()` and accessed through
+`get_active_compiler_options()`, however, it is advised that a user should use
+`finalize_compilation_options()` for getting the final options that should be used
+for optimization.
+"""
+
+
+def update_active_compiler_options(new_active_options: CompilerOptions) -> CompilerOptions:
+    """
+    Updates the set of active compiler options.
+
+    Merges the options passed as `new_active_options` with the currently active
+    compiler options. This set is used by `JaCeLowered.compile()` to determine
+    which options should be used.
+    The function will return the set of options that was active before the call.
+
+    To obtain the set of currently active options use `get_active_compiler_options()`.
+
+    Todo:
+        Make a proper context manager.
+    """
+    previous_active_options = _JACELOWERED_ACTIVE_COMPILE_OPTIONS.copy()
+    _JACELOWERED_ACTIVE_COMPILE_OPTIONS.update(new_active_options)
+    return previous_active_options
+
+
+def get_active_compiler_options() -> CompilerOptions:
+    """Returns the set of currently active compiler options."""
+    return _JACELOWERED_ACTIVE_COMPILE_OPTIONS.copy()
+
+
+def finalize_compilation_options(compiler_options: CompilerOptions | None) -> CompilerOptions:
+    """
+    Returns the final compilation options.
+
+    There are two different sources of optimization options. The first one is the global
+    set of currently active compiler options. The second one is the options that are
+    passed to this function, which takes precedence. Thus, the `compiler_options`
+    argument describes the difference from the currently active global options.
+
+    This function is used by `JaCeLowered` if it has to determine which options to use
+    for optimization, either for compiling the lowered SDFG or for computing the key.
+
+    Args:
+        compiler_options: The local compilation options.
+
+    See Also:
+        `get_active_compiler_options()` to inspect the set of currently active options
+        and `update_active_compiler_options()` to modify them.
+    """
+    return get_active_compiler_options() | (compiler_options or {})
```

Comment:
I do not understand your argument you can not call `finalize_compilation_options()` without argument as is argument does not have a default.

Your argument "The `update_active_compiler_options()` is not used" is not an argument.
The same applies to `jace.jit` which is not used inside the code base and by your argument can be removed, followed by that we should remove `JaCeWrapped`, I think you get my point.

The `update_active_compiler_options()` function is used inside the tests to disable the optimizations.

Furthermore, I will not give direct access to a global variable.
This will simply not happen!
However, you are welcomed to propose better names.

Reply:
Ok, I don't want to start a philosophical discussion here but please remember that in Python:
	-  _global_ variables are just attributes of a module object
	- and thus all classes and functions defined in a module are also global variables
	- and your code is giving direct read access to all of them because it's the only way to use them
	- and since in pure Python there is no way to forbid access, your code is also giving write access to them and any user can replace the definitions of functions and classes in JaCe modules and access underscored variables and do anything
	- and thus direct access to global variables is a common pattern in Python and it's just impossible to avoid in general, so context handlers are a much better way to deal with non-local state.   

Once this context is clear both of us, let's discuss the three functions handling compiler options:
- `update_active_compiler_options()` does not fix the actual problem of this approach which is that any user can change the default options at any point in an uncontrolled way. The proper way to do this is to use a context handler to make sure the defaults are changed only within a limited scope. If a context manager is not implemented, this function is 99.9% as dangerous as letting the user change the global variable directly herself, or even more, since this is now public API and may mislead users into thinking this is a safe and appropriate way to go, which it's not, so this function just should not exist.
- `get_active_compiler_options()` and `make_final_compilation_options()` heavily overlap since passing `None` as an argument to `make_final_compilation_options()` returns exactly the same value as `get_active_compiler_options()`. If you really want a function to access a global dictionary and/or merge it with another dictionary, then just create a simple function:

```python
def get_compiler_options(custom_options: CompilerOptions | None = None) -> CompilerOptions:
    return _JACELOWERED_ACTIVE_COMPILE_OPTIONS | (custom_options or {})
```

Reply:
I merged the two functions together and turned the update function into a context manager.